File: codes/java/82.删除排序链表中的重复元素-ii.java

class Solution {
	public ListNode deleteDuplicates(ListNode head) {
		if (head == null) {
			return null;
		}
		ListNode dummy = new ListNode(Integer.MIN_VALUE);
		ListNode prev = dummy;
		int duplicateVal = Integer.MIN_VALUE;
		while (head != null) {
			// 探查一下后继节点是否和当前节点重复, 如果重复则将值记录下来
			if (head.next != null && head.next.val == head.val) {
				duplicateVal = head.val;
			}
			if (head.val != duplicateVal) {
				prev.next = new ListNode(head.val);
				prev = prev.next;
			}
			head = head.next;
		}
		return dummy.next;
	}
}

File: codes/python/82.删除排序链表中的重复元素-ii.py

class Solution:
	def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
		if not head:
			return None

		dummy = ListNode(-1)
		prev = dummy
		# 设置<-100的值来作为哨兵
		duplicate_val = -1000

		while head:
			if head.next and head.next.val == head.val:
				duplicate_val = head.val
			if head.val != duplicate_val:
				prev.next = ListNode(head.val)
				prev = prev.next
			head = head.next

		return dummy.next